### ISSUE TITLE
Add new data types for symbolic remote data

### ISSUE THREAD

**Issue by sphuber** [aiidateam/aiida-core#4626]
This issue has flown from #2150 that is adding functionality to _stash_ files from the `remote_folder` of a `CalcJob` once it has terminated. This data would be attached as the `stash_folder` output node to the calculation job node. The question arose what `Data` type it should use. To be able to distinguish it from the `remote_folder ` it was suggested to use a different class from `RemoteData`, for example `StashData`.

This raised another question of how the subclassing should be done, because in a sense the `StashData` functions exactly as a `RemoteData`. They are both "symbolic links", in the sense that AiiDA only stores a reference to where the data that they contain is stored. If the remote data is removed AiiDA doesn't necessarily know but it can no longer be used for reproducibility. So far, `RemoteData` is the only data type that is symbolic in this way, but `StashData` would also become one. In that sense, it would make sense that they share a common parent class.

If we would not have to worry about backwards compatibility, we could have created a `SymbolicData` class that would be the parent of `RemoteData` and `StashData`. However, this would require a change in entry point string for `RemoteData`, since the `QueryBuilder` subclassing is based on that. It should be changed from `remote` to `symbolic.remote` if `symbolic` is the entry point name of the `SymbolicData` base class.

Instead, we can bombard `RemoteData` to the parent class, such that the entry point does not have to be changed. We then create two new subclasses `RemoteFolderData` and `RemoteStashData`:
```
"remote = aiida.orm.nodes.data.remote:RemoteData",
"remote.folder = aiida.orm.nodes.data.remote.folder:RemoteFolderData",
"remote.stash = aiida.orm.nodes.data.remote.stash:RemoteStashData",
```
Making it `RemoteFolderData` is actually quite good as it makes for a better parallel with the `FolderData` local analog.

The migration plan would be to already change the `remote_folder` output from `RemoteData` to `RemoteFolderData`. Queries for `RemoteData` will continue to work, since `RemoteFolderData` is a subclass, _unless_ users explicitly say `subclassing=False` in the query, which should be rare. We can then optionally add deprecation warnings when a raw `RemoteData` is stored and tell people to use the new `RemoteFolderData` instead. When sufficient time has passed for transition, we remove the warnings and make `RemoteData` not storable, essentially making it a pure base class (similar to `ProcessNode`). This step is optional, I am not yet sure if this is necessary and whether it is ok to allow `RemoteData` nodes to continue to be created. Actually not even sure if this is being done by users. If we do decide to make `RemoteData` non-storable, we can add a migration changing existing nodes' type strings from `remote` to `remote.folder`.

**Comment by ltalirz** [comment]
Just mentioning the related question of files vs folders https://github.com/aiidateam/aiida-core/issues/1609 (currently, there is no way to specify a "remote file").

One downside of the naming scheme described above is that the name still doesn't express that it represents a folder (all three entry points mentioned above actually represent folders).

If we want to eventually support remote files, we should probably think about this now - otherwise we can also decide that there isn't going to be support for remote files.


One suggestion from my side would be to replace the name `RemoteFolderData` with `RemoteScratchData`.
It's really the `RemoteData` that should be called `RemoteFolderData` - the differentiation on the inheritance level below it is between folders on "scratch" space and data on "stash" space.

**Comment by sphuber** [comment]
Thanks for the input @ltalirz . I did bring up the issue you link in our discussion, but we though that with the new hierarchy that we propose, this could easily be added later on in a backward compatible way once the need for a `RemoteFileData` arises. I don't want to really add things that we don't currently have a use case for.

> One downside of the naming scheme described above is that the name still doesn't express that it represents a folder (all three entry points mentioned above actually represent folders).

This is not fully true, is it? The `RemoteData` _shouldn't_ include the term folder since it doesn't necessarily have to be a folder, but merely be data that is on a remote. Only `RemoteStashData` does not explicitly include folder, but I am not convinced that it should. It can also just be a single file that is stashed. Should we have a separate type for this, i.e. `RemoteStashFileData`? Not sure if this is necessary.

> One suggestion from my side would be to replace the name RemoteFolderData with RemoteScratchData.
It's really the RemoteData that should be called RemoteFolderData - the differentiation on the inheritance level below it is between folders on "scratch" space and folders on "stash" space.

I don't agree that the differentiation between the two subclasses is based on their location on the remote. The `RemoteData` is currently used by `CalcJob`s as their working directory and is typically located on the scratch space, yes, but that is not it's only potential use. One could use a `RemoteFolderData` to also store it on a non-scratch space. The _real_ difference between `RemoteFolderData` and `RemoteStashData` is that the latter _can_ be something else than a normal folder on a filesystem and so might not directly be accessible like a `RemoteFolderData` is.

**Comment by ltalirz** [comment]
> This is not fully true, is it? The `RemoteData` _shouldn't_ include the term folder since it doesn't necessarily have to be a folder, but merely be data that is on a remote.

Ah sorry, perhaps I misunderstood.
Are you saying that the folder-specific methods in `RemoteData` will be deprecated/removed?
I.e. `listdir`, `listdir_withattributes`, `getfile` (with relpath) will move outside this class?

**Comment by sphuber** [comment]
Yes, anything that will be folder specific will then only be added to `RemoteFolderData`. The `RemoteData` will at some point merely become a base class and I am thinking we should even not allow instances of it to be stored.

**Comment by ltalirz** [comment]
Sounds good!

Just a final question: the `RemoteStashData` will represent a single file or a folder? If a folder, won't this require duplicating some if the functions added in `RemoteFolderData`?

**Comment by sphuber** [comment]
> Sounds good!
> 
> Just a final question: the `RemoteStashData` will represent a single file or a folder? If a folder, won't this require duplicating some of the functions added in `RemoteFolderData`?

No, `RemoteStashData` will not explicitly be a folder, because it can become a tarball. Or it can even represent files stored on tape. However, even if it _were_ just a folder, you could solve the problem of code duplication by extracting the folder logic to a mixin class, like `FolderDataMixin` which you then mix into the data plugins. That mixin does not necessarily have to be part of the entry point hierarchy, if you see what I mean.
